File: src/main/java/com/ultreon/mods/advanceddebug/client/menu/DebugGui.java
package com.ultreon.mods.advanceddebug.client.menu;

import com.mojang.blaze3d.vertex.PoseStack;
import com.ultreon.mods.advanceddebug.AdvancedDebug;
import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext;
import com.ultreon.mods.advanceddebug.api.client.menu.DebugPage;
import com.ultreon.mods.advanceddebug.api.client.menu.Formatter;
import com.ultreon.mods.advanceddebug.api.client.menu.IDebugGui;
import com.ultreon.mods.advanceddebug.api.common.*;
import com.ultreon.mods.advanceddebug.api.events.IInitPagesEvent;
import com.ultreon.mods.advanceddebug.client.Config;
import com.ultreon.mods.advanceddebug.client.formatter.FormatterContext;
import com.ultreon.mods.advanceddebug.client.input.KeyBindingList;
import com.ultreon.mods.advanceddebug.client.menu.pages.DefaultPage;
import com.ultreon.mods.advanceddebug.client.registry.FormatterRegistry;
import com.ultreon.mods.advanceddebug.text.ComponentBuilder;
import com.ultreon.mods.advanceddebug.util.InputUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.client.event.InputEvent;
import net.minecraftforge.client.gui.ForgeIngameGui;
import net.minecraftforge.client.gui.IIngameOverlay;
import org.lwjgl.glfw.GLFW;
import org.slf4j.Marker;
import org.slf4j.MarkerFactory;

import javax.annotation.Nullable;
import java.awt.*;
import java.util.List;
import java.util.*;

import static net.minecraft.ChatFormatting.*;
import static net.minecraft.util.FastColor.ARGB32.*;

/**
 * Client listener
 */
@SuppressWarnings("unused")
public final class DebugGui implements IIngameOverlay, IDebugGui {
    private static final DebugGui INSTANCE = new DebugGui();
    private static final List<DebugPage> pages = new ArrayList<>();
    private static final DebugPage DEFAULT = new DefaultPage();
    private static final FormatterRegistry FORMATTER_REGISTRY = FormatterRegistry.get();
    private static final Formatter<Object> DEFAULT_FORMATTER = new Formatter<>(Object.class, AdvancedDebug.res("object")) {
        @Override
        public void format(Object obj, IFormatterContext context) {
            Class<?> c = obj.getClass();

            context.classValue(c);
            context.identifier("@");
            context.hexValue(obj.hashCode());
        }
    };
    private static final Map<ResourceLocation, DebugPage> PAGE_REGISTRY = new HashMap<>();
    private static final Marker MARKER = MarkerFactory.getMarker("DebugGui");
    private Font font;
    private int page = 0;

    private DebugGui() {
        
    }

    @Override
    public void render(ForgeIngameGui gui, PoseStack pose, float partialTicks, int width, int height) {
        if (Minecraft.getInstance().options.renderDebug) {
            return;
        }

        font = Minecraft.getInstance().font;

        DebugPage debugPage = getDebugPage();
        Minecraft mc = Minecraft.getInstance();

        double scale = mc.getWindow().getGuiScale();
        double preferredScale = Config.USE_CUSTOM_SCALE.get() ? Config.CUSTOM_SCALE.get() : scale;
        Boolean useCustomScale = Config.USE_CUSTOM_SCALE.get();

        int rescaledWidth = (int) (((double) width * scale) / preferredScale);
        int rescaledHeight = (int) (((double) height * scale) / preferredScale);

        DebugRenderContext context = new DebugRenderContext(pose, rescaledWidth, rescaledHeight) {
            @Override
            protected void drawLine(PoseStack pose, Component text, int x, int y) {
                DebugGui.this.drawLine(pose, text, x, y);
            }
        };

        if (debugPage != null) {
            pose.pushPose();
            {
                pose.scale((float) ((1 * preferredScale) / scale), (float) ((1 * preferredScale) / scale), 1);
                ResourceLocation resourceLocation = debugPage.registryName();
                try {
                    if (Config.SHOW_CURRENT_PAGE.get()) {
                        drawLine(pose, new TextComponent("Debug Page: " + resourceLocation.toString()), 6, height - 16);
                    }
                    debugPage.render(pose, context);
                } catch (Exception e) {
                    if (resourceLocation != null)
                        AdvancedDebug.LOGGER.error(MARKER, "Error rendering debug page {}", resourceLocation, e);
                    else
                        AdvancedDebug.LOGGER.error(MARKER, "Error rendering debug page", e);

                    try {
                        if (resourceLocation != null)
                            context.left(new TextComponent("Error rendering debug page " + resourceLocation).withStyle(RED));
                        else
                            context.left(new TextComponent("Error rendering debug page").withStyle(RED));
                    } catch (Exception e1) {
                        AdvancedDebug.LOGGER.error(MARKER, "Error showing error on debug page", e1);
                    }
                }
            }
            pose.popPose();
        } else {
            DEFAULT.render(pose, context);
        }
    }

    private void drawLine(PoseStack pose, Component text, int x, int y) {
        Screen.fill(pose, x, y, x + font.width(text) + 2, (y - 1) + font.lineHeight + 2, 0x5f000000);
        font.draw(pose, text, x + 1, y + 1, 0xffffff);
    }

    public static DebugGui get() {
        return INSTANCE;
    }

    public List<DebugPage> getPages() {
        return Collections.unmodifiableList(pages);
    }

    @Override
    @SuppressWarnings("UnusedReturnValue")
    public <T extends DebugPage> T registerPage(T page) {
        pages.add(page);
        if (PAGE_REGISTRY.containsKey(page.registryName())) {
            throw new IllegalArgumentException("Duplicate debug page registration: " + page.registryName());
        }
        PAGE_REGISTRY.put(page.registryName(), page);
        return page;
    }

    @Override
    @Nullable
    public DebugPage getDebugPage() {
        fixPage();
        if (page == 0) {
            return null;
        }
        return pages.get(page - 1);
    }

    private void fixPage() {
        page %= pages.size() + 1;
        if (page < 0) {
            page += pages.size() + 1;
        }
    }

    @Override
    public int getPage() {
        return page;
    }

    @Override
    public void setPage(int page) {
        this.page = page % (pages.size() + 1);
    }

    @Override
    public void setPage(DebugPage page) {
        if (!pages.contains(page)) {
            throw new IllegalArgumentException("Page not registered.");
        }
        this.page = pages.indexOf(page);
    }

    @Override
    public void next() {
        setPage(getPage() + 1);
    }

    @Override
    public void prev() {
        setPage(getPage() - 1);
    }

    public void onKeyReleased(InputEvent.KeyInputEvent event) {
        if (event.getAction() == GLFW.GLFW_RELEASE && event.getKey() == KeyBindingList.DEBUG_SCREEN.getKey().getValue()) {
            if (InputUtils.isShiftDown()) prev();
            else next();
        }
    }

    @Deprecated(forRemoval = true)
    private IFormattable getFormatted(String s) {
        return new IFormattable() {
            @SuppressWarnings("removal")
            @Override
            public String toFormattedString() {
                return s;
            }
        };
    }

    private IFormattable getMultiplier(double multiplier) {
        return new Multiplier(multiplier);
    }

    private IFormattable getSize(int w, int h) {
        return new IntSize(w, h);
    }

    private IFormattable getSize(float w, float h) {
        return new FloatSize(w, h);
    }

    private IFormattable getPercentage(double value) {
        return new Percentage(value);
    }

    private Color getColor(Vec3 color) {
        return new Color((float) color.x, (float) color.y, (float) color.z);
    }

    private Color getColor(int rgb) {
        return new Color(red(rgb), green(rgb), blue(rgb), alpha(rgb));
    }

    private IFormattable getAngle(double angle) {
        return new Angle(angle * 360.0d);
    }

    private IFormattable getRadians(double angle) {
        return new Angle(Math.toDegrees(angle));
    }

    private IFormattable getDegrees(double angle) {
        return new Angle(angle);
    }

    public Component format(String text, Object obj, Object... objects) {
        StringBuilder sb = new StringBuilder();

        FormatterContext context = new FormatterContext();
        format(obj, context);

        ComponentBuilder builder = new ComponentBuilder();
        builder.append(text, WHITE);
        builder.append(": ", GRAY);
        builder.append(context.build());

        for (Object object : objects) {
            FormatterContext ctx = new FormatterContext();
            format(object, ctx);

            builder.append(", ", GRAY);
            builder.append(ctx.build());
        }

        return builder.build();
    }

    @Override
    @SuppressWarnings({"rawtypes", "unchecked"})
    public void format(Object obj, IFormatterContext context) {
        if (obj == null) {
            context.keyword("null");
        } else if (obj instanceof Class<?> c) {
            context.packageName(((Class<?>) obj).getPackage().getName() + ".");
            context.className(((Class<?>) obj).getSimpleName());
        } else {
            Formatter identified = FORMATTER_REGISTRY.identify(obj.getClass());
            if (identified != null) {
                identified.format(obj, context);
                return;
            }

            DEFAULT_FORMATTER.format(obj, context);
        }
    }

    @Override
    public Formatter<Object> getDefault() {
        return DEFAULT_FORMATTER;
    }

    public IInitPagesEvent createInitEvent() {
        return this::registerPage;
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/client/menu/DebugGui.kt
package com.ultreon.mods.advanceddebug.client.menu

import com.mojang.blaze3d.vertex.PoseStack
import com.ultreon.mods.advanceddebug.AdvancedDebug
import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext
import com.ultreon.mods.advanceddebug.api.client.menu.DebugPage
import com.ultreon.mods.advanceddebug.api.client.menu.Formatter
import com.ultreon.mods.advanceddebug.api.client.menu.IDebugGui
import com.ultreon.mods.advanceddebug.api.common.*
import com.ultreon.mods.advanceddebug.api.events.InitPagesEvent
import com.ultreon.mods.advanceddebug.client.Config
import com.ultreon.mods.advanceddebug.client.formatter.FormatterContext
import com.ultreon.mods.advanceddebug.client.input.KeyBindingList
import com.ultreon.mods.advanceddebug.client.menu.pages.DefaultPage
import com.ultreon.mods.advanceddebug.client.registry.FormatterRegistry
import com.ultreon.mods.advanceddebug.mixin.common.KeyMappingAccessor
import com.ultreon.mods.advanceddebug.text.ComponentBuilder
import com.ultreon.mods.advanceddebug.util.InputUtils
import com.ultreon.mods.advanceddebug.util.memoize
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiComponent
import net.minecraft.client.gui.components.Renderable
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import org.lwjgl.glfw.GLFW
import org.slf4j.MarkerFactory
import java.util.*

/**
 * Client listener
 */
@Suppress("unused", "UNUSED_VARIABLE", "UNUSED_PARAMETER")
class DebugGui private constructor() : GuiComponent(), Renderable, IDebugGui {
    private var font: Font? = null
    private val minecraft = Minecraft.getInstance()
    private var width = 0
    private var height = 0
    override fun render(poseStack: PoseStack, mouseX: Int, mouseY: Int, partialTick: Float) {
        if (Minecraft.getInstance().options.renderDebug) {
            return
        }
        updateSize()
        font = Minecraft.getInstance().font
        val debugPage = debugPage
        val mc = Minecraft.getInstance()
        val scale = mc.window.guiScale
        val preferredScale: Double = if (Config.useCustomScale.get()) Config.customScale.get()?.toDouble() ?: scale else scale
        val useCustomScale = Config.useCustomScale.get()
        val rescaledWidth = (width.toDouble() * scale / preferredScale).toInt()
        val rescaledHeight = (height.toDouble() * scale / preferredScale).toInt()
        val context: DebugRenderContext = object : DebugRenderContext(poseStack, rescaledWidth, rescaledHeight) {
            override fun drawLine(pose: PoseStack, text: Component, x: Int, y: Int) {
                this@DebugGui.drawLine(pose, text, x, y)
            }
        }
        if (debugPage != null) {
            poseStack.pushPose()
            run {
                poseStack.scale((1 * preferredScale / scale).toFloat(), (1 * preferredScale / scale).toFloat(), 1f)
                val resourceLocation = debugPage.registryName()
                try {
                    if (Config.showCurrentPage.get()) {
                        drawLine(poseStack, Component.literal("Debug Page: $resourceLocation"), 6, (height * (scale / preferredScale) - 16).toInt())
                    }
                    debugPage.render(poseStack, context)
                } catch (e: Exception) {
                    AdvancedDebug.logger.error(
                        marker,
                        "Error rendering debug page {}",
                        resourceLocation,
                        e
                    )
                    try {
                        context.left(
                            Component.literal("Error rendering debug page $resourceLocation").withStyle(
                                ChatFormatting.RED
                            )
                        )
                    } catch (e1: Exception) {
                        AdvancedDebug.logger.error(marker, "Error showing error on debug page", e1)
                    }
                }
            }
            poseStack.popPose()
        } else {
            Companion.default.render(poseStack, context)
        }
    }

    private fun updateSize() {
        width = minecraft.window.guiScaledWidth
        height = minecraft.window.guiScaledHeight
    }

    private fun drawLine(pose: PoseStack, text: Component, x: Int, y: Int) {
        fill(pose, x, y, x + font!!.width(text) + 2, y - 1 + font!!.lineHeight + 2, 0x5f000000)
        font!!.draw(pose, text, (x + 1).toFloat(), (y + 1).toFloat(), 0xffffff)
    }

    val pages: List<DebugPage>
        get() = Collections.unmodifiableList(Companion.pages)

    override fun <T : DebugPage> registerPage(page: T): T {
        Companion.pages.add(page)
        require(!pageRegistry.containsKey(page.registryName())) { "Duplicate debug page registration: " + page.registryName() }
        pageRegistry[page.registryName()] = page
        return page
    }

    override val debugPage: DebugPage?
        get() {
            fixPage()
            return if (page == 0) {
                null
            } else Companion.pages[page - 1]
        }

    private fun fixPage() {
        page %= Companion.pages.size + 1
        if (page < 0) {
            page += Companion.pages.size + 1
        }
    }

    override var page = 0
        set(value) {
            field = value % (Companion.pages.size + 1)
        }

    override fun setPage(page: DebugPage) {
        require(Companion.pages.contains(page)) { "Page not registered." }
        this.page = Companion.pages.indexOf(page)
    }

    override fun next() {
        page++
    }

    override fun prev() {
        page--
    }

    fun onKeyReleased(keyCode: Int, scanCode: Int, action: Int, modifiers: Int): Boolean {
        if (action == GLFW.GLFW_RELEASE && keyCode == (KeyBindingList.DEBUG_SCREEN as KeyMappingAccessor).key.value) {
            if (InputUtils.isShiftDown) prev() else next()
            return true
        }
        return false
    }

    fun format(text: String, obj: Any?, vararg objects: Any?): Component {
        val sb = StringBuilder()
        val context = FormatterContext()
        format(obj, context)
        val builder = ComponentBuilder()
        builder.append(text, ChatFormatting.WHITE)
        builder.append(": ", ChatFormatting.GRAY)
        builder.append(context.build())
        for (`object` in objects) {
            val ctx = FormatterContext()
            format(`object`, ctx)
            builder.append(", ", ChatFormatting.GRAY)
            builder.append(ctx.build())
        }
        return builder.build()
    }

    override fun format(obj: Any?, context: IFormatterContext) {
        when (obj) {
            null -> {
                context.keyword("null")
            }
            is Class<*> -> {
                context.packageName(obj.getPackage().name + ".")
                context.className(obj.simpleName)
            }
            else -> {
                val identified: Formatter<*>? = formatterRegistry.identify(obj.javaClass)
                if (identified != null) {
                    formatUnsafe(identified, obj, context)
                    return
                }
                defaultFormatter.format(obj, context)
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : Any> formatUnsafe(formatter: Formatter<T>, obj: Any, context: IFormatterContext) {
        formatter.format(obj as T, context)
    }

    override val default: Formatter<Any>
        get() {
            return defaultFormatter
        }

    fun createInitEvent(): InitPagesEvent {
        return object : InitPagesEvent {
            override fun <T : DebugPage> register(page: T): T {
                return registerPage(page)
            }
        }
    }

    companion object {
        private val instance = DebugGui()
        private val pages: MutableList<DebugPage> = ArrayList()
        val default: DebugPage by memoize { DefaultPage() }

        private val formatterRegistry = FormatterRegistry.get()
        private val defaultFormatter: Formatter<Any> = object : Formatter<Any>(
            Any::class, AdvancedDebug.res("object")
        ) {
            override fun format(obj: Any, context: IFormatterContext) {
                val c: Class<*> = obj.javaClass
                context.classValue(c)
                context.identifier("@")
                context.hexValue(obj.hashCode())
            }
        }
        private val pageRegistry: MutableMap<ResourceLocation, DebugPage> = HashMap()
        private val marker = MarkerFactory.getMarker("DebugGui")
        @JvmStatic
        fun get(): DebugGui {
            return instance
        }
    }
}